sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamBundleSource.java

```diff
@@ -0,0 +1,349 @@
+
+    // Values used for progress reporting.
+    private double fractionOfStreamBundleConsumed;
+
+    private double progressAtResponseStart;
+    private double progressAtResponseEnd;
+    private long rowsConsumedFromCurrentResponse;
+    private long totalRowsInCurrentResponse;
+
+    private @Nullable TableReference tableReference;
+    private @Nullable ServiceCallMetric serviceCallMetric;
+
+    private BigQueryStorageStreamBundleReader(
+        BigQueryStorageStreamBundleSource<T> source, BigQueryOptions options) throws IOException {
+      super(source);
+      this.source = source;
+      this.reader = BigQueryStorageReaderFactory.getReader(source.readSession);
+      this.parseFn = source.parseFn;
+      this.storageClient = source.bqServices.getStorageClient(options);
+      this.tableSchema = fromJsonString(source.jsonTableSchema, TableSchema.class);
+      this.currentStreamBundleIndex = 0;
+      this.fractionOfStreamBundleConsumed = 0d;
+      this.progressAtResponseStart = 0d;
+      this.progressAtResponseEnd = 0d;
+      this.rowsConsumedFromCurrentResponse = 0L;
+      this.totalRowsInCurrentResponse = 0L;
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    protected long getCurrentOffset() throws NoSuchElementException {
+      return currentStreamBundleIndex;
+    }
+
+    @Override
+    protected boolean isAtSplitPoint() throws NoSuchElementException {
+      if (currentStreamOffset == 0) {
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public boolean startImpl() throws IOException {
+      return readNextStream();
+    }
+
+    @Override
+    public boolean advanceImpl() throws IOException {
+      Preconditions.checkStateNotNull(responseIterator);
+      currentStreamOffset++;
+      return readNextRecord();
+    }
+
+    private synchronized boolean readNextStream() throws IOException {
+      BigQueryStorageStreamBundleSource<T> source = getCurrentSource();
+      if (currentStreamBundleIndex == source.streamBundle.size()) {
+        fractionOfStreamBundleConsumed = 1d;
+        return false;
+      }
+      ReadRowsRequest request =
+          ReadRowsRequest.newBuilder()
+              .setReadStream(source.streamBundle.get(currentStreamBundleIndex).getName())
+              .setOffset(currentStreamOffset)
+              .build();
+      tableReference = BigQueryUtils.toTableReference(source.readSession.getTable());
+      serviceCallMetric = BigQueryUtils.readCallMetric(tableReference);
+      LOG.info(
+          "Started BigQuery Storage API read from stream {}.",
+          source.streamBundle.get(currentStreamBundleIndex).getName());
+      responseStream = storageClient.readRows(request, source.readSession.getTable());
+      responseIterator = responseStream.iterator();
+      return readNextRecord();
```

Comment:
Since this method is now `synchronized`, the `ReadRows` call performed in the `readNextRecord` call will now be done with the object lock held. We need to find a way to avoid this.